R/cv_varsel_formula.R

```diff
@@ -0,0 +1,370 @@
+  	list(refmodel = refmod, p_sel = p_sel, p_pred = p_pred,
+         mu_test = mu_test, dis = refmod$dis, w_test = refmod$wsample,
+         msg = msg, y_test = refmod$y[test_points], test_points = test_points)
+  }, refmodels_cv, k_fold$test_points, msgs, SIMPLIFY = F)
+
+  ## Perform the selection for each of the K folds
+  if (verbose) {
+    print('Performing selection for each fold..')
+    pb <- utils::txtProgressBar(min = 0, max = K, style = 3, initial=0)
+  }
+  spath_cv <- lapply(seq_along(list_cv), function(fold_index) {
+    fold <- list_cv[[fold_index]]
+    family_kl <- fold$refmodel$family
+    out <- select_poc(method, fold$p_sel, fold$refmodel, family_kl, intercept,
+                      nv_max, penalty, verbose, opt, groups=groups)
+    if (verbose)
+      utils::setTxtProgressBar(pb, fold_index)
+    out
+  })
+
+  vind_cv <- lapply(spath_cv, function(e) e$vind)
+  if (verbose)
+    close(pb)
+
+  ## Construct submodel projections for each fold
+  as.search <- relax
+  if (verbose && !as.search) {
+    print('Computing projections..')
+    pb <- utils::txtProgressBar(min = 0, max = K, style = 3, initial=0)
+  }
+  p_sub_cv <- mapply(function(spath, fold_index) {
+    fold <- list_cv[[fold_index]]
+    family_kl <- fold$refmodel$family
+    vind <- spath$vind
+    p_sub <- .get_submodels_poc(spath, c(0, seq_along(vind)), family_kl,
+                                fold$p_pred, fold$refmodel, intercept,
+                                opt$regul, as.search=as.search)
+    if (verbose && !as.search)
+      utils::setTxtProgressBar(pb, fold_index)
+    return(p_sub)
+  }, spath_cv, seq_along(list_cv), SIMPLIFY = F)
+  if (verbose && !as.search)
+    close(pb)
+
+
+  # Helper function extract and combine mu and lppd from K lists with each
+  # n/K of the elements to one list with n elements
+  hf <- function(x) as.list(do.call(rbind, x))
+
+  ## Apply some magic to manipulate the structure of the list so that instead of
+  ## list with K sub_summaries each containing n/K mu:s and lppd:s, we have only
+  ## one sub_summary-list that contains with all n mu:s and lppd:s.
+  sub <- apply(
+    mapply(function(p_sub, fold) {
+      family_kl <- fold$refmodel$family
+      lapply(.get_sub_summaries_poc(p_sub, fold$test_points, fold$refmodel, family_kl),
+             data.frame)
+    }, p_sub_cv, list_cv),
+    1, hf)
+
+  ref <- hf(lapply(list_cv, function(fold) {
+    family_kl <- fold$refmodel$family
+    test <- fold$test_points
+    d <- list(y=fold$y_test)
+    data.frame(.weighted_summary_means_poc(d, family_kl, fold$mu_test, fold$dis))}))
+
+  # Combine also the K separate test data sets into one list
+  # with n y's and weights's.
+  d_cv <- hf(lapply(list_cv, function(fold) {
+    test <- fold$test_points
+    data.frame(y=fold$y_test)}))
+
+  list(vind_cv = vind_cv,
+       summaries = list(sub = sub, ref = ref),
+       d_test = c(d_cv, type = 'kfold'))
+}
+
+
+.get_kfold_poc <- function(refmodel, K, verbose, seed) {
```

Comment:
So, originally, the folds can come from two places:

1. either because the user has fitted K number of fits and passed that already, or,
2. because the user has passed a cvfits function to recover the fit of each fold from the original fit.

Here we should just retrieve the proper fit and compute the statistics so there's not much overhead involved. Maybe cvfits is more complicated if it requires fitting each time. But, as I said in a previous comment, I think this is deep enough for a next PR and don't mingle too much on this one?